linkerd/app/core/src/disco_cache.rs

```diff
@@ -0,0 +1,155 @@
+use futures::TryFutureExt;
+use linkerd_error::Error;
+use linkerd_idle_cache::{Cached, NewIdleCached};
+use linkerd_stack::{
+    layer, queue, CloneParam, FutureService, MapErrBoxed, NewQueueWithoutTimeout, NewService,
+    Oneshot, Param, QueueWithoutTimeout, Service, ServiceExt, ThunkClone,
+};
+use std::{fmt, hash::Hash, task, time};
+
+/// A [`NewService`] that extracts a `K`-typed key from each target to build a
+/// [`Cached`]<[`DiscoverThunk`]>.
+#[derive(Clone)]
+pub struct NewCachedDiscover<K, D, N>
+where
+    K: Clone + fmt::Debug + Eq + Hash + Send + Sync + 'static,
+    D: Service<K, Error = Error> + Clone + Send + Sync + 'static,
+    D::Response: Clone + Send + Sync,
+    D::Future: Send + Unpin,
+{
+    // NewService<K, Service<(), D::Response>>
+    cache: NewIdleCached<K, NewQueueThunk<NewDiscoverThunk<D>>>,
+
+    // NewService<D::Response>
+    inner: N,
+}
+
+/// The future that drives discovery to build an new inner service wrapped
+/// in the [`Cached`] decorator from the discovery lookup.
```

Comment:
nit, maybe worth making this slightly more explicit:
```suggestion
/// in the [`Cached`] decorator from the discovery lookup, so that the lookup is not
/// considered idle while services built from it exist.
```
or something?